api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java

```diff
@@ -211,7 +211,12 @@ private Expression bindInOperation(BoundTerm<T> boundTerm) {
         case IN:
           return new BoundLiteralPredicate<>(Operation.EQ, boundTerm, Iterables.get(convertedLiterals, 0));
         case NOT_IN:
-          return new BoundLiteralPredicate<>(Operation.NOT_EQ, boundTerm, Iterables.get(convertedLiterals, 0));
+          if (boundTerm.ref().field().isRequired()) {
+            // 'col NOT IN (1)' is equivalent to 'col != 1' iff the column has no nulls
```

Comment:
`null NOT IN (1)` in SQL evaluates to `null`. Similarly, `null != 1` evaluates to `null`. Since both are `null` and both end up being `false`, aren't these expressions equivalent even if there are nulls?

It seems like this is trying to work around the fact that `null != 1` evaluates to `true` for Iceberg expressions because Iceberg expressions don't use 3-value boolean logic.